### PR TITLE
make `RegExpExecArray` always include index 0

### LINKED ISSUE #50520: `RegExpExecArray` should have explicit always-defined properties
\[L]ooking a little closely at the ECMAScript spec, I do think that there is a mismatch between what we have and what gets created in the runtime. [`RegExpBuiltinExec` says that the following are created unconditionally](https://tc39.es/ecma262/#sec-regexpbuiltinexec):

* an `index` property
* an `input` property
* a property at index `0`
* a `groups` property that *is always set*, but may be `undefined`

So `RegExpExecArray` should look something kind of like

```ts
// src/lib/es5.d.ts

interface RegExpExecArray extends Array<string> {
    // always present
    index: number;

    // always present
    input: string;

    // always present - helpful for '--noUncheckedIndexedAccess'
    0: string;
}
```

```ts
// src/lib/es2018.regexp.d.ts

interface RegExpExecArray extends Array<string> {
    // always present - helpful for '--exactOptionalPropertyTypes'
    groups: { [key: string]: string } | undefined;
}
```

*Derived from content posted by @DanielRosenwasser in https://github.com/microsoft/TypeScript/issues/50211#issuecomment-1210008233*

______

In additionl to the above, it *seems* like `RegExpMatchArray`/`RegExpExecArray` are sometimes both returned from functions that undergo the same internal code paths. So we should probably ensure that one is compatible with the other, and have a test that ensures as much.

### PR DESCRIPTION
**Pull request by DetachHead**
Fixes #50520

- [ ] i don't know how the fourslash tests work or how to debug them. i'll take a closer look tomorrow
- [ ] should `RegExpMatchArray` be deleted and `RegExpExecArray` be used for both `match` and `exec` as suggested by https://github.com/microsoft/TypeScript/issues/50520#issuecomment-1242878018?

### CODE REVIEW

**Review by KotlinIsland — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
I think it would be better not to introduce a new interface just to share properties.

Maybe there is a valid question as to whether `RegExpMatchArray` should be an alias for `[string, ...string[]] | RegExpExecArray`. I think I'd rather keep it as-is to be honest.

**Comment by DanielRosenwasser** [comment]
You don't need a fourslash test - fourslash is for testing editor-specific functionality. A test in `tests/cases/compiler` for using these types and trying to access the properties on them is good.

```ts
// @strict: true,false
// @noUncheckedIndexedAccess: true
// @exactOptionalPropertyTypes: true

export function foo(matchResult: RegExpExecArray, execResult: RegExpExecArray) {
    matchResult[0].length;
    matchResult[999].length;
    matchResult.index + 0;
    matchResult.input.length;
    matchResult.groups["someVariable"].length;
    matchResult.groups = undefined;

    execResult[0].length;
    execResult[999].length;
    execResult.index + 0;
    execResult.input.length;
    execResult.groups["someVariable"].length;
    execResult.groups = undefined;

    if (Math.random()) {
        matchResult = execResult;
    }
    else {
        execResult = matchResult
    }
}
```

Many of these will produce errors, and that's expected.

**Comment by graphemecluster** [comment]
> i don't know how the fourslash tests work or how to debug them.

Whenever you add anything globally, including an interface, in the lib files, you have to add an entry in `src\harness\fourslashInterfaceImpl.ts`. But as suggested by Mr. Rosenwasser, you are not encouraged to introduce a new interface just to share properties.

> should `RegExpMatchArray` be deleted?

My suggestion would be removing the `0` property in `RegExpMatchArray` as well as making it deprecated instead of deleting the whole interface so as not to breaking existing code.

Overall, my suggested changes would be:

```ts
interface String {
    // ...
    match(regexp: string | RegExp): [string, ...string[]] | RegExpExecArray | null;
    // ...
}

// ...

/** @deprecated Use `[string, ...string[]] | RegExpExecArray` instead */
interface RegExpMatchArray extends Array<string> {
    /**
     * The index of the search at which the result was found.
     */
    index?: number;
    /**
     * A copy of the search string.
     */
    input?: string;
}

interface RegExpExecArray extends Array<string> {
    /**
     * The index of the search at which the result was found.
     */
    index: number;
    /**
     * A copy of the search string.
     */
    input: string;
    /**
     * The first match. This will always be present because `null` will be returned if there are no matches.
     */
    0: string;
}
```

And in `es2018.regexp.d.ts`:
```ts
interface RegExpMatchArray {
    /**
     * An object containing the values of named capturing groups with the keys being the corresponding group names,
     * or undefined if no named capturing groups are defined.
     */
    groups?: { [key: string]: string } | undefined;
}

interface RegExpExecArray {
    /**
     * An object containing the values of named capturing groups with the keys being the corresponding group names,
     * or undefined if no named capturing groups are defined.
     */
    groups: { [key: string]: string } | undefined;
}

// (Add a newline before EOF)
```

**Comment by DetachHead** [comment]
> I think it would be better not to introduce a new interface just to share properties.

i'll remove that then. however would you mind elaborating on why that's a bad idea? i've noticed several instances of duplicated properties that would be better off extending a shared interface imo (eg. #46781) so i'm curious as to why duplicating code is the preferred approach?

> You don't need a fourslash test - fourslash is for testing editor-specific functionality.

ah ok. the reason i brought it up is because my change caused some of the fourslash tests to fail, according to @graphemecluster it was because of that new interface i added

**Comment by graphemecluster** [comment]
I don’t mean to remove the addition of `0` to `RegExpExecArray` (but only to `RegExpMatchArray`), I’ve edited the comment to clarify that. Sorry!
Anyway I seems to care to much about compatibility; let’s leave it to @DanielRosenwasser.

**Comment by DanielRosenwasser** [comment]
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt

File: tests/baselines/reference/regexpExecAndMatchTypeUsages(strict=true).errors.txt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+tests/cases/compiler/regexpExecAndMatchTypeUsages.ts(6,5): error TS2532: Object is possibly 'undefined'.
+tests/cases/compiler/regexpExecAndMatchTypeUsages.ts(7,5): error TS2412: Type 'undefined' is not assignable to type '{ [key: string]: string; }' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the type of the target.
+tests/cases/compiler/regexpExecAndMatchTypeUsages.ts(10,5): error TS2532: Object is possibly 'undefined'.
+tests/cases/compiler/regexpExecAndMatchTypeUsages.ts(13,5): error TS2532: Object is possibly 'undefined'.
+tests/cases/compiler/regexpExecAndMatchTypeUsages.ts(13,5): error TS2532: Object is possibly 'undefined'.
+tests/cases/compiler/regexpExecAndMatchTypeUsages.ts(14,5): error TS2412: Type 'undefined' is not assignable to type '{ [key: string]: string; }' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the type of the target.
+
+
+==== tests/cases/compiler/regexpExecAndMatchTypeUsages.ts (8 errors) ====
+    export function foo(matchResult: RegExpExecArray, execResult: RegExpExecArray) {
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/regexpExecAndMatchTypeUsages(strict=true).errors.txt:12]: dang, this was supposed to be a match array

**Pull request merged**